kafka-plugins-client/src/main/java/io/cdap/plugin/sink/KafkaOutputFormat.java

```diff
@@ -43,85 +43,84 @@ public class KafkaOutputFormat extends OutputFormat<Text, Text> {
   private KafkaProducer<String, String> producer;
 
   @Override
-  public void checkOutputSpecs(JobContext jobContext) throws IOException, InterruptedException {
+  public void checkOutputSpecs(JobContext jobContext) {
   }
 
   @Override
-  public OutputCommitter getOutputCommitter(TaskAttemptContext taskAttemptContext) throws IOException,
-    InterruptedException {
+  public OutputCommitter getOutputCommitter(TaskAttemptContext taskAttemptContext) {
     return new OutputCommitter() {
       @Override
-      public void setupJob(JobContext jobContext) throws IOException {
+      public void setupJob(JobContext jobContext) {
         // no-op
       }
 
       @Override
-      public void setupTask(TaskAttemptContext taskContext) throws IOException {
+      public void setupTask(TaskAttemptContext taskContext) {
         //no-op
       }
 
       @Override
-      public boolean needsTaskCommit(TaskAttemptContext taskContext) throws IOException {
+      public boolean needsTaskCommit(TaskAttemptContext taskContext) {
         return false;
       }
 
       @Override
-      public void commitTask(TaskAttemptContext taskContext) throws IOException {
+      public void commitTask(TaskAttemptContext taskContext) {
         //no-op
       }
 
       @Override
-      public void abortTask(TaskAttemptContext taskContext) throws IOException {
+      public void abortTask(TaskAttemptContext taskContext) {
         //no-op
       }
     };
   }
 
   @Override
-  public RecordWriter<Text, Text> getRecordWriter(TaskAttemptContext context)
-    throws IOException, InterruptedException {
+  public RecordWriter<Text, Text> getRecordWriter(TaskAttemptContext context) {
     Configuration configuration = context.getConfiguration();
 
     // Extract the topics
     String topic = configuration.get("topic");
 
-    Properties props = new Properties();
-    // Configure the properties for kafka.
-    props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG,
+    // CDAP-9178: cached the producer object to avoid being created on every batch interval
+    if (producer == null) {
+      Properties props = new Properties();
+      // Configure the properties for kafka.
+      props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG,
               configuration.get(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG));
-    props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
+      props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
               configuration.get(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG));
-    props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+      props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
               configuration.get(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG));
-    props.put("compression.type", configuration.get("compression.type"));
+      props.put("compression.type", configuration.get("compression.type"));
+      props.put(ProducerConfig.LINGER_MS_CONFIG, 100);
```

Comment:
Any particular reason for the value 100?

Reply:
There is a white paper by Confluence: https://www.confluent.io/blog/optimizing-apache-kafka-deployment/ that suggests linger.ms values should be between 10-100 ms for best performance. I tried different values 10,50,100 and noticed that it performed best with the value of 100 ms.

The number of records generated per 30 mins for different values were:
linger.ms 10ms- 133,243,283
linger.ms 50ms- 135,367,345
linger.ms 100ms- 142,434,822

Reply:
You can add a comment in the code explaining why you are using the value 100

Reply:
We should allow to configure it either from UI or from runtime arguments

Reply:
+1 to making it configurable. UI allows `Additional Kafka Producer Properties`, so maybe we can look for the property value in configuration and set it to a default value if not specified.

Reply:
Fixed.